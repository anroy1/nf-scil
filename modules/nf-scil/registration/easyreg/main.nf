

process REGISTRATION_EASYREG {
    tag "$meta.id"
    label 'process_single'

    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://scil.usherbrooke.ca/containers/scilus_2.0.2.sif':
        'scilus/scilus:2.0.2' }"

    input:
    tuple val(meta), path(reference), path(floating), path(ref_segmentation), path(flo_segmentation)

    output:
    tuple val(meta), path("${prefix}_reference_segmentation.nii.gz") , emit: ref_seg
    tuple val(meta), path("${prefix}_floating_segmentation.nii.gz")  , emit: flo_seg
    tuple val(meta), path("${prefix}_reference_registered.nii.gz")   , emit: ref_reg
    tuple val(meta), path("${prefix}_floating_registered.nii.gz")    , emit: flo_reg
    tuple val(meta), path("${prefix}_forward_field.nii.gz")          , emit: fwd_field
    tuple val(meta), path("${prefix}_backward_field.nii.gz")         , emit: bak_field
    path "versions.yml"                                              , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def prefix = task.ext.prefix ?: "${meta.id}"
    def threads = task.ext.threads ? "--threads " + task.ext.threads : ""
    def affine = task.ext.affine ? "--affine_only " : ""

    """
    if [[ -f "$ref_segmentation" ]];
    then
        reference_segmentation = $ref_segmentation
    else
        reference_segmentation = ${prefix}_reference_segmentation.nii.gz
    fi

        if [[ -f "$flo_segmentation" ]];
    then
        floating_segmentation = $flo_segmentation
    else
        floating_segmentation = ${prefix}_floating_segmentation.nii.gz
    fi

    mri_easyreg --ref $reference --flo $floating --ref_seg \${reference_segmentation} --flo_seg --ref_seg \${reference_segmentation} \
        --flo_reg ${prefix}_floating_registered.nii.gz --ref_reg ${prefix}_reference_registered.nii.gz --fwd_field ${prefix}_forward_field.nii.gz \
        --bak_field ${prefix}_backward_field.nii.gz $threads $affine

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        freesurfer: 7.4
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"

    """
    mri_easyreg -h

    touch ${prefix}_reference_segmentation.nii.gz
    touch ${prefix}_floating_segmentation.nii.gz
    touch ${prefix}_reference_registered.nii.gz
    touch ${prefix}_floating_registered.nii.gz
    touch ${prefix}_forward_field.nii.gz
    touch ${prefix}_backward_field.nii.gz

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        freesurfer: 7.4
    END_VERSIONS
    """
}
