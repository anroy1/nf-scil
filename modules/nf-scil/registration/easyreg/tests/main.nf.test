nextflow_process {

    name "Test Process REGISTRATION_EASYREG"
    script "../main.nf"
    process "REGISTRATION_EASYREG"

    tag "modules"
    tag "modules_nfcore"
    tag "registration"
    tag "registration/easyreg"

    tag "subworkflows"
    tag "subworkflows/load_test_data"

    setup {
            run("LOAD_TEST_DATA", alias: "LOAD_DATA") {
                script "../../../../../subworkflows/nf-scil/load_test_data/main.nf"
                process {
                    """
                    input[0] = Channel.from( [ "freesurfer.zip" ] )
                    input[1] = "test.load-test-data"
                    """
                }
            }
        }

    test("registration - easyreg") {
        config "./nextflow.config"
        when {
            process {
                """
                input[0] = LOAD_DATA.out.test_data_directory
                        .map{ test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/fa.nii.gz"),
                        file("\${test_data_directory}/t1.nii.gz"),
                        [],
                        []
                ]}
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    niftiMD5SUM(process.out.ref_seg.get(0).get(1), 6),
                    niftiMD5SUM(process.out.flo_seg.get(0).get(1), 6),
                    niftiMD5SUM(process.out.ref_reg.get(0).get(1), 6),
                    niftiMD5SUM(process.out.flo_reg.get(0).get(1), 6),
                    niftiMD5SUM(process.out.fwd_field.get(0).get(1), 6),
                    niftiMD5SUM(process.out.bak_field.get(0).get(1), 6),
                    process.out.versions
                ).match() }
            )
        }
    }
}
