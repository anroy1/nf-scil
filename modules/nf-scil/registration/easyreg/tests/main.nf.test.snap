{
    "registration - easyreg": {
        "content": [
            "test_reference_segmentation.nii.gz:md5:header,b3c282f86451ca606b428bd9ffcacd10,data,6b5ed67cf63a32966fee1ae179539053",
            "test_floating_segmentation.nii.gz:md5:header,ff063c025479983639fd545e1e8fd690,data,7e70ecf0bdb68271b0f9a0aa22711365",
            "test_reference_registered.nii.gz:md5:header,83b3e1eaaaa72c367ffcdd4cfa002406,data,fcbb08556a5d6950d79d661900c1b1c8",
            "test_floating_registered.nii.gz:md5:header,c5e41f89848f91c53a9a7be44970d4b1,data,236aff072c85b1405d1d3095f7a5267b",
            "test_forward_field.nii.gz:md5:header,74c92ee4cd3c4abfecf3da6cdb1d4650,data,2826360a14ad92253cea497bb2e5d055",
            "test_backward_field.nii.gz:md5:header,2184515d220e6c1eeb48adae2134ce7e,data,06b7b17a98829d7c75f3e1ebb77cfdc2",
            [
                "versions.yml:md5,4661210880c42a986923e8257f64c760"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0-rc1",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-11T14:02:22.638457"
    }
}